### DISCUSSION TITLE
Wind Speed Estimator in ROSCO

### DISCUSSION THREAD

**Discussion by mohsamaz** [OpenFAST/openfast#2573] [General]
Dear Nikhar,
I have a question regarding your paper titled "A Reference Open-Source Controller for Fixed and Floating Offshore Wind Turbine."

In Figure 6 of the paper, you show a wind speed estimator. However, it appears that the wind speed used in this figure is not real but rather generated with an average of 11 m/s. In contrast, real wind turbine operations would involve actual wind speed measurements with significantly more fluctuation.

Could you please explain this? Specifically, was the wind speed estimator shown in Figure 6 based on idealized conditions, and how does it compare to real-world scenarios where fluctuating wind speeds are present?

Does wind speed stimator implemented in FORTRAN?
Is it possible to design a new wind speed stimator with python?

**Comment by dzalkind** [comment]
Hi,

Figure 6 shows the wind speed estimate compared to the rotor averaged wind speed.  The rotor average wind speed will smooth the signal, compared to a single point measurement in space.  The rotor average wind speed was calculated from a full-field turbulence generated using TurbSim.  I don't know the exact parameters used to generate the file, but I'm sure you could reproduce that figure with your own configuration.

The wind speed estimator was implemented in Fortran here: https://github.com/NREL/ROSCO/blob/122c75864c1c0c1a828975180f46e77220771e60/rosco/controller/src/ControllerBlocks.f90#L187

I'm sure it's possible to design a wind speed estimator in python.

Best, Dan

**Comment by mohsamaz** [comment]
Hi everyone,

Thank you @dzalkind for your earlier response!

I’m curious about the use of a wind speed estimator. What is the primary reason for using one? When a turbine operates under real conditions, wind speeds can vary widely, ranging from 0 to 20 m/s. How does such an estimator perform under these varying real-world conditions?

Additionally, I’m interested in designing a custom wind speed estimator in Python and replacing the existing estimators in ROSCO. Could anyone provide insights or guidance on how to approach this?

Thank you in advance for your help!

Best regards,

**Reply by dzalkind** [reply]
Wind speed estimators are used because point measurements, like an anemometer on the nacelle, are noisy and can have interference.

The wind speed estimator should work under varying real world conditions.

I would conduct a literature review and review existing implementations, like the one I provided, and implement the estimator in python with some kind of turbine model.  The ROSCO toolbox has a simple simulator that is demonstrated here: https://github.com/NREL/ROSCO/blob/122c75864c1c0c1a828975180f46e77220771e60/Examples/04_simple_sim.py#L24

**Comment by mohsamaz** [comment]
I would like to estimate wind speed as shown in Figure 6 of your paper, "A Reference Open-Source Controller for Fixed and Floating Offshore Wind Turbine."

I created three DISCON files based on the IEA-15-240-RWT-UMaineSemi with the following cases:
WE_Mode:0 
WE_Mode:1
WE_Mode:2

I used case_gen to generate .fst files with the following parameters:

p["TMax"]=600
p["HydroFile|WaveMod"]=0
p["InflowFile|WindType"] = 2 
p["InflowFile|Filename_Uni"] = '../_WIND/steps_9_12.dat'
p["ServoFile|DLL_FileName"] = "libdiscon_ROSCO29.dll"
p["EDFile|BlPitch(1)"] = 0
p["EDFile|BlPitch(2)"] = 0
p["EDFile|BlPitch(3)"] = 0
p['EDFile|RotSpeed'] = 8.5
p['HydroFile|WaveTMax'] = 0 

I ran simulations for each .fst file with the wind speed stepped from 9 to 12 m/s. The simulation results are shown below:


![output_plot](https://github.com/user-attachments/assets/1ee341f9-7c3c-4751-93cf-d4b7bcc6de6d)


My Question:
Can I get the wind speed estimate from my simulations? If so, how can I enable or activate the wind speed estimator to include it as an output?

Thank you for your time and insights!

**Reply by dzalkind** [reply]
Hi,

Wind speed estimates can be output from the ROSCO library, when `LoggingLevel` > 0 in the DISCON.in and if you evaluate the `WE_Vw` channel in the `.dbg` output file. 

We don't recommend using `WE_Mode` of 1.

Best, Dan

**Comment by mohsamaz** [comment]
Dear @dzalkind,

I attempted to replicate Figure 6 from your paper "A Reference Open-Source Controller for Fixed and Floating Offshore Wind Turbine."with the following parameters:

Wind speed turbulence: 8 m/s
Wind turbine: 5 MW
WE_Mode: 2

turbSim TimeStep = 0.01
Simulation DT = 0.00625

Could you please confirm if I plotted it correctly? Upon closer inspection, the estimation plot appears slightly delayed and doesn’t align perfectly with the averaged wind speed.

Is this behavior expected, or is there something I need to adjust?
Is there any standard for set the time step of turbSim and timestep of simulation?

Best regards,
![Figure 2025-01-27 131600](https://github.com/user-attachments/assets/b47b65d5-cd9d-48ff-bf37-950ea6c540bd)
![Figure 2025-01-27 133038](https://github.com/user-attachments/assets/d472458a-5216-4ed6-93f3-baa7e2b27413)

**Reply by dzalkind** [reply]
These plots look like expected wind speed estimator performance.  There will be a slight delay because the estimate is based off of measurements.  The standard time step I use in TurbSim is 0.05 s.  In OpenFAST, it depends on your solver choices and the dynamics of your model.  When working with a new model, I typically sweep the time step to find the highest, but safe, choice.

**Comment by mohsamaz** [comment]
Dear @dzalkind,

First of all thank you for all helps.
The ROSCO toolbox has a simple simulator.
as I know simple simulator helps to run without use time consuming openfast simulation and it is the purpose of this example.
However, is it also possible to in this example(04_simple_sim.py) also use openfast instead of simple model (sim class)?
If so, could you please provide some guidance or suggestions on how to do that?

Best regards,

**Reply by dzalkind** [reply]
Hi, 

Here are some relevant examples: https://github.com/NREL/ROSCO/blob/main/Examples/05_openfast_sim.py

The `run_FAST_ROSCO` class incorporates many more features, as well: https://github.com/NREL/ROSCO/blob/cfda6af95cda15616b840199c24fbd208e63893c/Examples/27_power_ref_control.py#L59

I hope this helps.

Best, Dan

**Comment by mohsamaz** [comment]
Hi @dzalkind,
thank you for your answer.
I want to clarify,
Is it also possible to use ws_series and ws_wd_series but instead of 1dof model use openfast?
best regards

**Reply by dzalkind** [reply]
Not directly, but you can make wind inputs suitable for InflowWind that are similar.

**Comment by mohsamaz** [comment]
Dear @dzalkind,
I run simulation of NREL 5MW 
0.0125   DT 
InflowFile|WindType: 3,with wind speed 11 m/s and TimeStep 0.05 .       
I compared the different WE_modes.      

WE_Mode=0         : One-second low pass filtered hub height wind speed,

![image](https://github.com/user-attachments/assets/4ff8946b-93ea-42a7-ab77-0deceb1dc772)

in this case(WE_Mode=0) the first order low pass filter of Wind1VelX is not consistent with one second low pass filter.
I assumed that this low pass filter is the same LPFilter in ROSCO.

WE_Mode=1:      Immersion and Invariance Estimator,
 
![image](https://github.com/user-attachments/assets/1c09b413-3fff-430a-bb40-f06f4bd1100c)

WE_Mode=2:       Extended Kalman Filter
![image](https://github.com/user-attachments/assets/3cebaabf-d31c-4c07-81ee-16dacdea33ba)

in case(WE_Mode=2) rmse=0.58m/s which in this case is different from the value of 0.48 m/s in your paper.

1-could you please have a look if my simulations is correct or I should correct something.

Best regards,

**Reply by dzalkind** [reply]
These results seem reasonable.  WE_Mode of 1 is not actively developed, so I would not recommend its use at this time.

Best, Dan

**Comment by mohsamaz** [comment]
Dear @dzalkind,
thank you for you recent answer.

could you please answer more specific about this questions?

in this case(WE_Mode=0) the first order low pass filter of Wind1VelX is not consistent with one second low pass filter.
I assumed that this low pass filter is the same LPFilter in ROSCO fortran filter file.
1-Why first order low pass filter of Wind1VelX and WE_Vw (WE_mode 0) are not exactly same?

2-in case(WE_Mode=2) rmse=0.58m/s which in this case is different from the value of 0.48 m/s that you mentioned in your paper[A reference open-source controller for fixed and floating offshore wind turbines, Figure 6]

when you mentioned rsme is 0.48 m/s, you wanted to show that this accuracy is achieavable in every simulation or not?
thank you in advance for your help.
Best regards,

**Reply by dzalkind** [reply]
1. It doesn't look like `WE_Vw` is a one second low pass filter. It's likely a higher time constant.
2. Different wind files will have different errors.  Reproducing the results exactly like you suggest is not achievable.

**Comment by mohsamaz** [comment]
Dear @dzalkind,
I am sharing a python code that computes low-pass filter of Wind1VelX and compare plot with WE_Vw (WE_Mode=0).
https://github.com/mohsamaz/openfast/blob/main/test_LPfilter.py
Howeve, I can not in WE_mode0 by filtering the Wind1VelX get the VE_Vw.
In my code If I increase  'cornering_frequency_Hz' value to a large value(1000 Hz) I get the orignal ,but if I I set 'cornering_frequency_Hz' to 1Hz,  it is not the VE_Vw.

it would be appritiated if you help me to find out what is the issue.
Best regards,

**Reply by dzalkind** [reply]
I'm sorry, but without more information and figures, I can not help you figure out the issue.  I should note that issues with ROSCO would probably be better addressed in the issues of that repository.

**Comment by mohsamaz** [comment]
Hello @dzalkind,
thank you for you prompt reply,

I am adding some figures.
cornering_frequency = 1Hz:
![image](https://github.com/user-attachments/assets/f98347c0-7019-4d27-aef9-ae6314013ab0)

cornering_frequency =5Hz:
![image](https://github.com/user-attachments/assets/7cb1d0c2-5fc3-4f56-8670-f2dacf25656b)

cornering_frequency =100Hz:
![image](https://github.com/user-attachments/assets/35e988a4-de14-4f31-b0a7-389b91cc6947)

also I alreasy add data(Run.out and Run.RO.dbg) and code here:
https://github.com/mohsamaz/openfast

For the next time I ask my question in Rosco repository.

Best regards

**Reply by dzalkind** [reply]
The filter on the wind speed measurement in ROSCO is far slower than 1Hz, the default is F_WECornerFreq 0.20944 **rad/s**.

**Comment by mohsamaz** [comment]
Thank you again ,
I assumed 1Hz because In ROSCO controllerBlock.f90 line 189 mentions:
WE_Mode = 0, Filter hub height wind speed as passed from servodyn using first order low pass filter with 1Hz cornering frequency.

by the way, according to your kind guidance if
0.20944 rad/s is 0.0333Hz, then my figure is this:

![image](https://github.com/user-attachments/assets/78ef8bc7-dfc8-43d4-95eb-a1f14f6e6075)

Best regards,
